kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -148,8 +141,54 @@ maybeTermEquals
     -> TermLike variable
     -> TermLike variable
     -> MaybeT unifier (Pattern variable)
-maybeTermEquals notSimplifier =
-    maybeTransformTerm (equalsFunctions notSimplifier)
+maybeTermEquals notSimplifier childTransformers first second =
+    Builtin.Int.unifyInt first second
+    <|> Builtin.Bool.unifyBool first second
+    <|> Builtin.String.unifyString first second
+    <|> unifyDomainValue first second
+    <|> unifyStringLiteral first second
+    <|> equalAndEquals first second
+    <|> bytesDifferent first second
+    <|> bottomTermEquals SideCondition.topTODO first second
+    <|> termBottomEquals SideCondition.topTODO first second
+    <|> variableFunctionEquals first second
+    <|> variableFunctionEquals second first
+    <|> equalInjectiveHeadsAndEquals childTransformers first second
+    <|> sortInjectionAndEquals childTransformers first second
+    <|> constructorSortInjectionAndEquals first second
+    <|> constructorAndEqualsAssumesDifferentHeads first second
+    <|> overloadedConstructorSortInjectionAndEquals
+            childTransformers
+            first
+            second
+    <|> Builtin.Bool.unifyBoolAnd childTransformers first second
+    <|> Builtin.Bool.unifyBoolOr childTransformers first second
+    <|> Builtin.Bool.unifyBoolNot childTransformers first second
+    <|> Builtin.Int.unifyIntEq childTransformers notSimplifier first second
+    <|> Builtin.String.unifyStringEq childTransformers notSimplifier first second
+    <|> Builtin.KEqual.unifyKequalsEq
+            childTransformers
+            notSimplifier
+            first
+            second
+    <|> Builtin.Endianness.unifyEquals first second
+    <|> Builtin.Signedness.unifyEquals first second
+    <|> unifyDefinedModifier
+            (Builtin.Map.unifyEquals childTransformers)
+            first
+            second
+    <|> Builtin.Map.unifyNotInKeys childTransformers notSimplifier first second
+    <|> unifyDefinedModifier
+            (Builtin.Set.unifyEquals childTransformers)
+            first
+            second
+    <|> Builtin.List.unifyEquals
+            SimplificationType.Equals
+            childTransformers
+            first
+            second
+    <|> domainValueAndConstructorErrors first second
+    <|> unifyDefined childTransformers first second
```

Comment:
One thing we should try here is to use `ExceptT` instead of `MaybeT`, like this:

```suggestion
maybeTermEquals notSimplifier childTransformers first second = run $ do
    hoist $ Builtin.Int.unifyInt first second
    hoist $ Builtin.Bool.unifyBool first second
    hoist $ Builtin.String.unifyString first second
    hoist $ unifyDomainValue first second
    hoist $ unifyStringLiteral first second
    hoist $ equalAndEquals first second
    hoist $ bytesDifferent first second
    hoist $ bottomTermEquals SideCondition.topTODO first second
    hoist $ termBottomEquals SideCondition.topTODO first second
    hoist $ variableFunctionEquals first second
    hoist $ variableFunctionEquals second first
    hoist $ equalInjectiveHeadsAndEquals childTransformers first second
    hoist $ sortInjectionAndEquals childTransformers first second
    hoist $ constructorSortInjectionAndEquals first second
    hoist $ constructorAndEqualsAssumesDifferentHeads first second
    hoist $ overloadedConstructorSortInjectionAndEquals childTransformers first second
    hoist $ Builtin.Bool.unifyBoolAnd childTransformers first second
    hoist $ Builtin.Bool.unifyBoolOr childTransformers first second
    hoist $ Builtin.Bool.unifyBoolNot childTransformers first second
    hoist $ Builtin.Int.unifyIntEq childTransformers notSimplifier first second
    hoist $ Builtin.String.unifyStringEq childTransformers notSimplifier first second
    hoist $ Builtin.KEqual.unifyKequalsEq childTransformers notSimplifier first second
    hoist $ Builtin.Endianness.unifyEquals first second
    hoist $ Builtin.Signedness.unifyEquals first second
    hoist $ unifyDefinedModifier (Builtin.Map.unifyEquals childTransformers) first second
    hoist $ Builtin.Map.unifyNotInKeys childTransformers notSimplifier first second
    hoist $ unifyDefinedModifier (Builtin.Set.unifyEquals childTransformers) first second
    hoist $ Builtin.List.unifyEquals SimplificationType.Equals childTransformers first second
    hoist $ domainValueAndConstructorErrors first second
    hoist $ unifyDefined childTransformers first second
  where
    hoist :: Monad m => MaybeT m a -> ExceptT a m ()
    hoist m =
        lift (runMaybeT m) >>= \case
            Nothing -> return ()
            Just a -> throwE a
    {-# INLINE hoist #-}
    run m =
        lift (runExceptT m) >>= \case
            Left a -> return a
            Right () -> empty
```

`runExceptT`, `ExceptT`, and `throwE` come from `Control.Error`. I haven't had time to try profiling this change, but the Core it generates is very different. There is still a big set of nested `let` statements, but they appear outside the `>>=`s so it's like `let let let >>= >>= >>=` instead of `>>= let >>= let >>= let`. This may improve the generated code. (Or it may not.)

Reply:
Just so that this is documented here as well, this suggestion turned out to decrease performance.